orangecontrib/prototypes/widgets/contingency_table.py

```diff
@@ -0,0 +1,365 @@
+        elif j == 1 or not self.circles and j == m - 1:
+            selection = QItemSelection(index(i, 2), index(i, m - 1))
+
+        if selection is not None:
+            self.selectionModel().select(
+                selection, QItemSelectionModel.ClearAndSelect)
+
+    def _item(self, i, j):
+        return self.tablemodel.item(i, j) or QStandardItem()
+
+    def _set_item(self, i, j, item):
+        self.tablemodel.setItem(i, j, item)
+
+    def set_variables(self, variablev, variableh, **kwargs):
+        """
+        Sets class headers and top headers and initializes table structure.
+
+        Parameters
+        ----------
+        variablev : Orange.data.variable.DiscreteVariable
+            Class headers are set to ``variablev.values``, top header is set to ``variablev.name``.
+        variableh : Orange.data.variable.DiscreteVariable
+            Class headers are set to ``variableh.values``, top header is set to ``variableh.name``.
+        """
+        self.classesv = variablev.values
+        self.classesh = variableh.values
+        self.headerv = variablev.name
+        self.headerh = variableh.name
+        self.initialize(**kwargs)
+
+    def set_headers(self, classesv, classesh, headerv=None, headerh=None, **kwargs):
+        """
+        Sets class headers and top headers and initializes table structure.
+
+        Parameters
+        ----------
+        classesv : :obj:`list` of :obj:`str`
+            Vertical class headers.
+        classesh : :obj:`list` of :obj:`str`
+            Horizontal class headers.
+        headerv : :obj:`str`, optional
+            Vertical top header.
+        headerh : :obj:`str`, optional
+            Horizontal top header.
+        """
+        self.classesv = classesv
+        self.classesh = classesh
+        self.headerv = headerv
+        self.headerh = headerh
+        self.initialize(**kwargs)
+
+    def initialize(self, circles=False):
+        """
+        Initializes table structure. Class headers must be set beforehand.
+
+        Parameters
+        ----------
+        circles : :obj:`bool`, optional
+            Turns on circle display. All table values should be between 0 and 1 (inclusive).
+        """
+        assert self.classesv is not None and self.classesh is not None
+
+        self.circles = circles
+
+        if self.circles:
+            self.setItemDelegate(CircleItemDelegate(Qt.white))
+        else:
+            self.setItemDelegate(BorderedItemDelegate(Qt.white))
+        item = self._item(0, 2)
+        item.setData(self.headerh, Qt.DisplayRole)
+        item.setTextAlignment(Qt.AlignCenter)
+        item.setFlags(Qt.NoItemFlags)
+
+        self._set_item(0, 2, item)
+        item = self._item(2, 0)
+        item.setData(self.headerv, Qt.DisplayRole)
+        item.setTextAlignment(Qt.AlignHCenter | Qt.AlignBottom)
+        item.setFlags(Qt.NoItemFlags)
+        self.setItemDelegateForColumn(0, gui.VerticalItemDelegate())
+        self._set_item(2, 0, item)
```

Comment:
could you split this method into two or more? (one to create cell items, one for headers) As it is, it is hard to follow what is going on.